ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -717,10 +762,26 @@ func (r *RayJobReconciler) checkActiveDeadlineAndUpdateStatusIfNeeded(ctx contex
 	if rayJob.Spec.ActiveDeadlineSeconds == nil || time.Now().Before(rayJob.Status.StartTime.Add(time.Duration(*rayJob.Spec.ActiveDeadlineSeconds)*time.Second)) {
 		return false
 	}
-	logger.Info("The RayJob has passed the activeDeadlineSeconds. Transition the status to `Failed`.", "StartTime", rayJob.Status.StartTime, "ActiveDeadlineSeconds", *rayJob.Spec.ActiveDeadlineSeconds)
-	rayJob.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusFailed
+
 	rayJob.Status.Reason = rayv1.DeadlineExceeded
 	rayJob.Status.Message = fmt.Sprintf("The RayJob has passed the activeDeadlineSeconds. StartTime: %v. ActiveDeadlineSeconds: %d", rayJob.Status.StartTime, *rayJob.Spec.ActiveDeadlineSeconds)
+
+	failedCount := int32(0)
+	if rayJob.Status.Failed != nil {
+		failedCount = *rayJob.Status.Failed
+	}
+	failedCount++
+	rayJob.Status.Failed = ptr.To[int32](failedCount)
+
+	if rayJob.Spec.BackoffLimit != nil && *rayJob.Spec.BackoffLimit > 0 &&
```

Comment:
There's 3 different places where we check for retry now:
1. checkK8sJobAndUpdateStatusIfNeeded
2. checkActiveDeadlineAndUpdateStatusIfNeeded
3. Reconcile -- inside rayv1.JobDeploymentStatusRunning case

I'd like to refactor this code a bit so that we only need to check for retry in one place in a future PR

Reply:
I like that updateRayJobStatus doesn't have any logic for the RayJob state machine and only does equality check + the API call. I think adding the backoffLimit logic here would be more confusing in the end because we're pushing the statemachine logic outside the switch in Reconcile. 

I removed case 2) for `checkActiveDeadlineAndUpdateStatusIfNeeded` so there's only 2 places where backoffLimit is checked. I think this is fine for now. What do you think?

Reply:
I may prefer to add logic in `updateRayJobStatus`. It's fine to update the status in this function because it is the only place to update the CR status. Maintaining retry in a single place also enables us to easily understand when we should and should not retry. Sorry for the last-minute change.

Reply:
Agree it's better to keep it in a single place. How about a function like `checkBackoffLimt()` before `updateRayJobStatus()` that checks for retries and updates the deployment status? I had this in the initial version but moved it into the switch case

Reply:
> How about a function like checkBackoffLimt() before updateRayJobStatus() that checks for retries and updates the deployment status?

SGTM

> I had this in the initial version but moved it into the switch case

https://media2.giphy.com/media/HyCxJErg9jTCU/giphy.gif?cid=5a38a5a2ljk87nmglp89z45h6zdw6un6x8kzdl8dv7xushc4&ep=v1_gifs_search&rid=giphy.gif&ct=g

Reply:
![](https://media2.giphy.com/media/HyCxJErg9jTCU/giphy.gif?cid=5a38a5a2otamclxbl457xu8bf7gmol0tza91y4hv7qvngc5b&ep=v1_gifs_search&rid=giphy.gif&ct=g)

Reply:
finally figure it out

Reply:
haha nice :)

Reply:
Added new method `checkBackoffLimitAndUpdateStatusIfNeeded` to consolidate retry checks into a single place